clab/netlink.go

```diff
@@ -4,153 +4,183 @@ import (
 	"fmt"
 	"net"
 	"os"
-	"os/exec"
-	"strings"
-	"sync"
 
+	"github.com/containernetworking/plugins/pkg/ns"
 	"github.com/google/uuid"
 	log "github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
 )
 
-func (c *cLab) InitVirtualWiring() {
-	// list interfaces
-	log.Debug("listing system interfaces...")
-	interfaces, err := net.Interfaces()
-	if err != nil {
-		log.Warnf("failed to get system interfaces:%v", err)
-		return
-	}
-	log.Debugf("found %d interfaces", len(interfaces))
-	for i := range interfaces {
-		if strings.HasPrefix(interfaces[i].Name, "clab-") {
-			log.Debugf("deleting interface %s", interfaces[i].Name)
-			l, err := netlink.LinkByName(interfaces[i].Name)
-			if err != nil {
-				log.Debugf("failed to find interface for deletion by name: %v", interfaces[i].Name)
-				continue
-			}
-			err = netlink.LinkDel(l)
-			if err != nil {
-				log.Debugf("failed to delete interface %s: %v", interfaces[i].Name, err)
-			}
-		}
-	}
+type vEthEndpoint struct {
+	Link     *netlink.Link
+	LinkName string
+	NSName   string // netns name
+	NSPath   string // netns path
+	Bridge   string // bridge name of veth is connected to it
+	MTU      int    // endpoint MTU
 }
 
 // CreateVirtualWiring provides the virtual topology between the containers
 func (c *cLab) CreateVirtualWiring(link *Link) (err error) {
-	log.Infof("Create virtual wire : %s:%s <--> %s:%s", link.A.Node.LongName, link.A.EndpointName, link.B.Node.LongName, link.B.EndpointName)
-	if link.A.Node.Kind != "bridge" && link.B.Node.Kind != "bridge" {
-		return c.createAToBveth(link)
-	}
-	return c.createvethToBridge(link)
+	log.Infof("Create virtual wire: %s:%s <--> %s:%s", link.A.Node.LongName, link.A.EndpointName, link.B.Node.LongName, link.B.EndpointName)
+	return c.createVethWiring(link)
 }
 
-func (c *cLab) createAToBveth(l *Link) error {
-	interfaceA := fmt.Sprintf("clab-%s", genIfName())
-	interfaceB := fmt.Sprintf("clab-%s", genIfName())
-
-	cmd := exec.Command("sudo", "ip", "link", "add", interfaceA, "type", "veth", "peer", "name", interfaceB)
-	err := runCmd(cmd)
+// createVethWiring connects containers (or container and bridge) using veth pair
+// using information contained within l Link
+func (c *cLab) createVethWiring(l *Link) error {
```

Comment:
I had an idea to add vxlan tunnels later (to interconnect containers/bridges over the network) and/or different link types like macvlan. That will be driven by link type property, which is now only veth.

But since it is not going to happen soon, I can merge this two funcs now and split them later once the changes will become more materialized

Reply:
done in 583d1ec